### PR TITLE
feat: Migrate to next js

### PR DESCRIPTION
**Pull request by chef-jojo**
This PR only does covert CRA, React Router to Next.js.
preview: https://pcs-nextjs.netlify.app/

Things need to do before merge:
1. rename all env prefix to `NEXT_PUBLIC_` 
2. update uikit to support SSR https://github.com/pancakeswap/pancake-toolkit/pull/370 https://github.com/pancakeswap/pancake-toolkit/pull/371
3. sentry integration: add env `SENTRY_AUTH_TOKEN`

### CODE REVIEW

File: src/pages/nfts/profile/[accountAddress]/activity.tsx [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import React from 'react'
+import { NftMarketLayout } from 'views/Nft/market/Layout'
+import NftProfile from 'views/Nft/market/Profile'
+import ActivityHistory from 'views/Nft/market/Profile/components/ActivityHistory'
+import SubMenu from 'views/Nft/market/Profile/components/SubMenu'
+
+const NftProfileActitivyPage = () => {
```
Comment (memoyil) [inline src/pages/nfts/profile/[accountAddress]/activity.tsx:7]: Typo

File: src/pages/nfts/profile/[accountAddress]/achievements.tsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import { useAchievementsForAddress, useProfileForAddress } from 'state/profile/hooks'
+import { NftMarketLayout } from 'views/Nft/market/Layout'
+import NftProfile from 'views/Nft/market/Profile'
+import Achievements from 'views/Nft/market/Profile/components/Achievements'
+
+const NftProfileAchievementsPage = () => {
+  const accountAddress = useRouter().query.accountAddress as string
+  const { profile: profileHookState } = useProfileForAddress(accountAddress)
+  const { profile } = profileHookState || {}
+  const { achievements, isFetching: isAchievementFetching } = useAchievementsForAddress(accountAddress)
```
Comment (memoyil) [inline src/pages/nfts/profile/[accountAddress]/achievements.tsx:12]: Am I missing something or useAchievements for the connected account will not be used?

File: src/views/Nft/market/Profile/ConnectedProfile.tsx [hunk 3]
```diff
@@ -31,7 +25,7 @@ const TabMenuWrapper = styled(Box)`
   }
 `
 
-const ConnectedProfile = () => {
+const ConnectedProfile: FC = ({ children }) => {
```
Comment (memoyil) [inline src/views/Nft/market/Profile/ConnectedProfile.tsx:28]: As far as I understood, We will fetch achievements twice when connected user profile viewed right?

Comment (chef-jojo) [reply src/views/Nft/market/Profile/ConnectedProfile.tsx:28]: I will move it to `getLayout`, should preserve state and layout between the navigation.

File: src/pages/nfts/profile/[accountAddress]/achievements.tsx (hunk 2 above)
Comment (chef-jojo) [reply src/pages/nfts/profile/[accountAddress]/achievements.tsx:12]: I'm late for the party. so the only difference of connected vs unconnected here is with redux or not, right 🌚

Comment (memoyil) [reply src/pages/nfts/profile/[accountAddress]/achievements.tsx:12]: Yes, Unconnected uses hooks with ...ForAddress, Connected takes it from redux

Comment (chef-jojo) [reply src/pages/nfts/profile/[accountAddress]/achievements.tsx:12]: These two should eventually merge together. I will put connected/unconnected pattern back for now, try not to change too much in this PR.
About `useAchievementsForAddress` being used twice in `UnconnectedProfile` and `UnConnectedAchievement`
I can't think of any quick fix for this one.  Let's hope https://github.com/pancakeswap/pancake-frontend/pull/2965 will be merged soon

File: src/__tests__/utils/apr.test.ts [hunk 4]
```diff
@@ -45,6 +46,6 @@ describe('getFarmApr', () => {
       '0x0ed7e52944161450477ee417de9cd3a859b14fd0',
     )
     expect(cakeRewardsApr).toEqual(4204800)
-    expect(lpRewardsApr).toEqual(10.5)
+    expect(lpRewardsApr).toEqual(lpAprs['0x0ed7e52944161450477ee417de9cd3a859b14fd0'])
```
Comment (chef-jojo) [inline src/__tests__/utils/apr.test.ts:49]: why this test pass in the past 🤯

**Comment by BaptisteGarcin** [comment]
why nextjs? Isn't the point of the blockchain to not rely on a single hosted server?

File: src/pages/_app.tsx [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            (function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
+            new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
+            j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
+            'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
+            })(window,document,'script','dataLayer', '${process.env.NEXT_PUBLIC_GTAG}');
+          `,
+        }}
+      />
+      <Providers>
+        <PersistGate loading={<PageLoader />} persistor={persistor}>
```
Comment (chef-jojo) [inline src/pages/_app.tsx:84]: Remove this when we are fully ready for SSR

**Comment by chef-jojo** [comment]
> why nextjs? Isn't the point of the blockchain to not rely on a single hosted server?

A Frontend server (most likely serverless) to help rendering react doesn't mean to opposite blockchain. We can be more "decentralize" by serving more static files on IPFS https://github.com/pancakeswap/pancake-frontend/issues/2046
Moving to Nextjs can unleash more ability on FE, 
for example:
Farms and Pools are getting too big right now, https://github.com/pancakeswap/pancake-frontend/blob/develop/src/config/constants/pools.ts
we eventually need to have some cache and middleware to handle these stuff

**Comment by BaptisteGarcin** [comment]
> > why nextjs? Isn't the point of the blockchain to not rely on a single hosted server?
> 
> A Frontend server (most likely serverless) to help rendering react doesn't mean to opposite blockchain. We can be more "decentralize" by serving more static files on IPFS #2046 Moving to Nextjs can unleash more ability on FE, for example: Farms and Pools are getting too big right now, https://github.com/pancakeswap/pancake-frontend/blob/develop/src/config/constants/pools.ts we eventually need to have some cache and middleware to handle these stuff

thanks for answering! I still don't really understand why since ipfs can be pulled as well without nextjs, but no problem it's ok thanks for taking the time, really appreciated

**Comment by zoltan31** [comment]
> > > why nextjs? Isn't the point of the blockchain to not rely on a single hosted server?
> > 
> > 
> > A Frontend server (most likely serverless) to help rendering react doesn't mean to opposite blockchain. We can be more "decentralize" by serving more static files on IPFS #2046 Moving to Nextjs can unleash more ability on FE, for example: Farms and Pools are getting too big right now, https://github.com/pancakeswap/pancake-frontend/blob/develop/src/config/constants/pools.ts we eventually need to have some cache and middleware to handle these stuff
> 
> thanks for answering! I still don't really understand why since ipfs can be pulled as well without nextjs, but no problem it's ok thanks for taking the time, really appreciated

Switching to Next.js doesn't help nor harm decentralization since the current React app is (probably) hosted on a centralized cloud service which is then distributed through CDN.
Using IPFS to host static files can be a solution to become fully decentralized, but that would probably hurt the user experience.

File: src/views/Nft/market/Profile/ConnectedProfile.tsx (hunk 3 above)
Comment (chef-jojo) [reply src/views/Nft/market/Profile/ConnectedProfile.tsx:28]: fix fetch twice https://github.com/pancakeswap/pancake-frontend/pull/2978/commits/05605c0da344cb6ebdfdab8514bb979e435a738a

File: src/pages/_app.tsx [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            Conditionally rendering `PersistGate` will cause react dom hydrate error.
+            "Warning: Did not expect server HTML to contain a <div> in <div>."
+           */}
+          {process.browser ? (
+            <PersistGate loading={null} persistor={persistor}>
+              <App {...props} />
+            </PersistGate>
+          ) : (
+            <App {...props} />
+          )}
```
Comment (chef-jojo) [inline src/pages/_app.tsx:85]: In order to render HTML in first paint, we have to remove `PersistGate` at server side. It leads to a well known warning about react dom hydrate error. Mark it here, hope we can find a way to solve it later.

**Review by Chef-Cheems — APPROVED** [review]

**Review by chefilip — COMMENTED** [review]
lgtm

**Pull request merged**